### PR TITLE
[JitBackend] Set up stubs for JIT tensor backend

### PR DESCRIPTION
**Pull request by StrongerXi**
### Summary

This PR sets up the JIT tensor's implementation of the Flashlight Tensor API (i.e., TensorBase & TensorBackend).

Conceptually, this implementation serves as the "parser" from Tensor operations to the JIT IR, with the following components:

1. `JitBackend` is the parser for common operations
2. `JitTensorBase` is the one that actually holds the JIT IR node, and will deal with mutation/indexing/shallowCopy in the future.
3. `JitTensor<T>` serves as a template trait to turn any Tensor into a "jitted" Tensor, i.e., `JitTensorBase`. 

`JitTensor<T>` is the only component that's statically aware of the actual tensor backend type. For `JitTensorBase` and `JitBackend`, only the latter has access to a `TensorBackend wrappedBackend` reference.

The user interface for JIT is very simple and makes 0 assumption on the actual tensor backend:
```c++
// in general: fl::setDefaultTensorType<JitTensor<...>>();
fl::setDefaultTensorType<JitTensor<OneDnnTensor>>();
```

### Test Plan (required)
Unit tests are added.
```
make JitTensorTest
./flashlight/fl/test/JitTensorTest
```

### CODE REVIEW

File: flashlight/fl/test/CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,21 @@
...
+add_library(
+  jit_test_utils
+  ${DIR}/tensor/jit/Utils.cpp
+)
+set_target_properties(jit_test_utils PROPERTIES LINKER_LANGUAGE CXX)
+target_include_directories(
+  jit_test_utils
+  PUBLIC
+  ${PROJECT_SOURCE_DIR}
+)
```
Comment (StrongerXi) [inline flashlight/fl/test/CMakeLists.txt:15]: Need some guidance on best practices here...

File: flashlight/fl/tensor/backend/jit/JitBackend.cpp [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+
+namespace fl {
+
+JitBackend::JitBackend(
+    TensorBackend& wrappedBackend,
+    std::function<Tensor(Node*)> jitTensorCreator)
+    : wrappedBackend_(wrappedBackend), jitTensorCreator_(jitTensorCreator) {}
+
+TensorBackendType JitBackend::backendType() const {
+  // Implementers of a backend should create their own option in the
```
Comment (jacobkahn) [inline flashlight/fl/tensor/backend/jit/JitBackend.cpp:26]: nit: remove comment?

File: flashlight/fl/tensor/backend/jit/JitTensorBase.cpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+FL_JIT_TENSOR_ASSIGN_OP_STUB(inPlaceMultiply); // *=
+FL_JIT_TENSOR_ASSIGN_OP_STUB(inPlaceDivide); // /=
+#undef FL_JIT_TENSOR_ASSIGN_OP_TYPE
+#undef FL_JIT_TENSOR_ASSIGN_OP
+
+Node* JitTensorBase::node() const {
+  return node_;
+}
+
+JitTensorBase& toJitTensorBase(const Tensor& tensor) {
```
[the remark this answers is not in the document]
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensorBase.cpp:160]: Had to do this for `OneDnnTensor` (couldn't find the rationale anymore...).

Corrected for JitTensor -- will deal with it when it actually becomes an issue.

File: flashlight/fl/tensor/backend/jit/JitTensor.h [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  // 1 static instance per jitted T.
+  JitBackend& backend() const override {
+    auto creator = [](Node* node) { return toTensor<JitTensor>(node); };
+    static JitBackend backend(wrappedBackend(), creator);
+    return backend;
+  }
+
+  explicit JitTensor(Node* node) : JitTensorBase(node) {}
+
+  JitTensor() : JitTensor({0}, fl::dtype::f32, nullptr, Location::Host) {}
```
[the remark this answers is not in the document]
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensor.h:42]: Yeah I ran into the same thing for [OneDNN backend](https://github.com/flashlight/flashlight/blob/d5c70ac69109ec3cbb7ce1d11767c5cc68a2a017/flashlight/fl/tensor/backend/onednn/OneDnnTensor.cpp#L153-L154). Was wondering if we could omit the empty constructor here and just let the TensorCreator pass in these defaults to the other constructor. But anyways, documented.

File: flashlight/fl/tensor/backend/jit/JitTensor.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  TensorBackend& wrappedBackend() const override {
+    static TensorBackend& wrappedBackend = toTensor<T>().backend();
+    return wrappedBackend;
+  }
+
+ public:
+  // 1 static instance per jitted T.
+  JitBackend& backend() const override {
+    auto creator = [](Node* node) { return toTensor<JitTensor>(node); };
+    static JitBackend backend(wrappedBackend(), creator);
```
[the remark this answers is not in the document]
Comment (StrongerXi) [reply flashlight/fl/tensor/backend/jit/JitTensor.h:36]: Actually the linker is smart enough to figure out these `JitBackend` instances (from the same `JitTensor<T>`) in different compilation units are the same:) -- https://stackoverflow.com/questions/19366615/static-member-variable-in-class-template. Documented.

**Comment by StrongerXi** [comment]
Address all feedbacks.

**Pull request closed without merging**